### PR TITLE
Use the Alternate entry, if it exists, in ICCBased Colour Space dictionaries (issue 5836, issue 5939, issue 6055)

### LINKED ISSUE #5836: PdfJS rendering image picture COLOR wrongly
I just tested pdfJS using the code inside example folder (examples/learning/prevnext.html). Below is picture of the pdf (which is rendered to canvas) and the actual pdf image. As it can be seen, the color which originally  purple has turned become bright yellow colour. It also happen with some of other pdf pages (which turn light brown into red color). Is there any solution to avoid this?

Here is original pdf file which i used to test: https://danesrti8sy2q.cloudfront.net/public/issues/6492/pages/844aa31ba77b830038f5b556f2fdf48862023cfb/background.pdf

screen shot 2015-03-13 at 3 16 09 pm
screen shot 2015-03-13 at 3 16 25 pm

One more question, the file is stored in cloudfront, but the code will throw to error: Refused to get unsafe header "Accept-Ranges". I try to add some headers like Accept-Ranges, Content-Encoding, etc.. to the S3 bucket policy, but it doesn't help. Would appreciate if someone can help me figure out what need to be changed. A sample S3 bucket policy would be very helpful. Thanks.

### LINKED ISSUE #5939: Some text renders with incorrect colours
This file: https://www.usenix.org/system/files/login/articles/login_apr15_02_grosvenor_041315.pdf

    "PDF dccdf906ffde4b5a97d5ef7478b37c46 [1.6 Adobe PDF Library 11.0 / Adobe InDesign CC 2014 (Macintosh)] (PDF.js: 1.0.1040)"

Firefox 37.0

Incorrect in PDF.js:
15-04-2015 4-20-34 am

Correct in Adobe Acrobat XI:
15-04-2015 4-21-17 am

Also works correctly in Chrome 41's built-in viewer.

### LINKED ISSUE #6055: Text Missing / Colors Displayed Incorrectly
In the following PDF, http://aualum.org/pdfs/magazine/mediakit2015.pdf , text colored #10233e is being displayed as #ff0000. Text with the color #f27f3c is not displayed at all. How the PDF should appear and how it is being displayed are shown in the two screenshots below.

In Chrome:
screen shot 2015-05-27 at 9 10 30 am

In Firefox:
screen shot 2015-05-27 at 9 10 45 am

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Fixes #5836.
Fixes #5939.
Fixes #6055.

_The relevant part of the specification:_ http://www.adobe.com/content/dam/Adobe/en/devnet/acrobat/pdfs/PDF32000_2008.pdf#G7.3804856.

### CODE REVIEW

**Comment by timvandermeij** [comment]
Nice work! I couldn't quite figure out what was wrong, but I'm glad you did! One question though: the spec says, for the `Alternate` entry, that "non-conforming readers may use this colour space". It sounds like we should check if the color space is unsupported and then fall back to the alternate color space, whereas in your patch we directly use the alternate color space without further checking. Personally I think that is fine, as it fixes the problems and the alternate entry is not specified for nothing, but do you think we need to do more checking, if even possible?

**Comment by Snuffleupagus** [comment]
Good questions! I don't have very convincing answers, but this is my take on it:

> [...], whereas in your patch we directly use the alternate color space without further checking.

I also thought about that when trying to make sense of the specification, but I don't know what that should look like while also fixing the referenced issues.
Without actual test-cases it seems difficult to be sure how this should be implemented, since just trying to adhere to the specification has shown to be a bit of a problem in practice :-)

> [...], but do you think we need to do more checking, if even possible?

Ideally we probably should, but again without actual test-cases (i.e. PDF files found in the wild), that seems difficult to me.
Given that we currently have three example files that this patch fixes, and unfortunately quite poor test-coverage for this part of the code-base, I think I'd rather risk breaking some other PDF files with this change. At least that way, we might actually get test-cases to work with :-P

**Comment by timvandermeij** [comment]
Agreed, and since mupdf and poppler also work exactly this way (see the ninja edit in my previous comment), this should be fine. I'll review and test this once the Windows bot is alive again.

**Comment by Snuffleupagus** [comment]
Actually, based both on the specification and the implementation in `mupdf`, we should perhaps try to check that the number of components match? I'll check to see if that is simple to implement.

**Comment by Snuffleupagus** [comment]
I've just uploaded a new version, which checks that the components match and (indirectly) that it's not a `Pattern` colour space. Unfortunately this requires, as far as I can see, that we parse the CS which is somewhat unfortunate. However, in the referenced PDF files, this doesn't seem to affect the rendering times in any noticeable way.

/botio-linux preview

**Comment by Snuffleupagus** [comment]
> One question though: the spec says, for the Alternate entry, that "non-conforming readers may use this colour space" [...]

Actually, the complete quote is:
> (Optional) An alternate colour space that shall be used in case the one specified in the stream data is not supported. Non-conforming readers may use this colour space. [...]

Since we currently completely ignore the `contents` of the stream data, I think that just using the `/Alternate` CS, if it exists, is *exactly* the right thing to do here!
Having re-read the paragraph a few times, I think that we're now doing what the specification says here.

File: src/core/colorspace.js [hunk 1]
```diff
@@ -284,6 +284,17 @@ var ColorSpace = (function ColorSpaceClosure() {
           var stream = xref.fetchIfRef(cs[1]);
           var dict = stream.dict;
           numComps = dict.get('N');
+          alt = dict.get('Alternate');
+          if (alt) {
+            var altIR = ColorSpace.parseToIR(alt, xref, res);
+            // Parse the /Alternate CS to ensure that the number of components
+            // are correct, and that it is not a PatternCS.
```
Comment (Snuffleupagus) [inline src/core/colorspace.js]: I'm going to push a new version of the patch, which slightly improves this comment. No functional code changes are made, though!

**Comment by timvandermeij** [comment]
Thank you!

**Comment by Snuffleupagus** [comment]
@timvandermeij Thank you too!
Your questions prompted me to go back and *really* read the specification, which in my opinion significantly helped improve the quality of the final patch :-)

**Comment by Snuffleupagus** [comment]
@timvandermeij Thank you too!
Your questions prompted me to go back and *really* read the specification; so you helped to *significantly* improve the quality of the final patch :-)

**Pull request merged**